Bottom: 0158150512fb965c5d1cce7529c373f29e15cc38
Top:    0cef237f08958dc13312a8fb8cc29de2a949c6c7
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-02-08 01:59:20 +0000

Add a hacky name-mangling thingy for loading boards











---

diff --git a/src/create.c b/src/create.c
index fda2613..373c213 100644
--- a/src/create.c
+++ b/src/create.c
@@ -37,6 +37,7 @@
 #include <memory.h>
 #include <setjmp.h>
 #include <stdlib.h>
+#include <stdio.h>
 
 #include "global.h"
 
@@ -54,6 +55,8 @@
 #include "undo.h"
 #include "vendor.h"
 
+#include <glib.h>
+
 #ifdef HAVE_LIBDMALLOC
 #include <dmalloc.h>
 #endif
@@ -625,6 +628,57 @@ CreateNewPointInPolygon (PolygonTypePtr Polygon, LocationType X,
   return (point);
 }
 
+static char *refdes_map_file = "refdes.map";
+static GHashTable *refdes_map_table = NULL;
+
+void RefdesMapInit (void)
+{
+  FILE *mapfile;
+  char *from = NULL;
+  char *to = NULL;
+
+  refdes_map_table = g_hash_table_new_full (g_str_hash, g_str_equal,
+                                            g_free, g_free);
+
+  if (refdes_map_file == NULL)
+    return;
+
+  printf ("Refdes map file %s\n", refdes_map_file);
+
+  mapfile = fopen (refdes_map_file, "rb");
+
+  if (mapfile == NULL) {
+    fprintf(stderr, "Could not open [%s]\n", refdes_map_file);
+    return;
+  }
+
+  while (!feof(mapfile) && !ferror(mapfile)) {
+    free (from);
+    free (to);
+
+    from = NULL;
+    to = NULL;
+
+    fscanf (mapfile, "%ms %ms\n", &from, &to);
+
+    if (from == NULL || to == NULL)
+      continue;
+
+    g_hash_table_insert (refdes_map_table, g_strdup (from), g_strdup (to));
+
+    printf ("Stashing rename command: %s to %s\n", from, to);
+  }
+
+  free (to);
+  free (from);
+}
+
+
+char *RefdesMapMap (char *from)
+{
+  return g_hash_table_lookup (refdes_map_table, from);
+}
+
 /* ---------------------------------------------------------------------------
  * creates an new element
  * memory is allocated if needed
@@ -637,6 +691,19 @@ CreateNewElement (DataTypePtr Data, ElementTypePtr Element,
 		  LocationType TextX, LocationType TextY, BYTE Direction,
 		  int TextScale, FlagType TextFlags, Boolean uniqueName)
 {
+
+  /* HACK: Mapping between refdes */
+  if (NameOnPCB != NULL) {
+    char *to;
+    to = RefdesMapMap (NameOnPCB);
+    if (to != NULL) {
+//      free (NameOnPCB);
+      printf ("Mapping name from %s to %s\n", NameOnPCB, to);
+      NameOnPCB = malloc (sizeof (char) * (strlen (to) + 1));
+      strcpy (NameOnPCB, to);
+    }
+  }
+
   if (!Element)
     Element = GetElementMemory (Data);
 
diff --git a/src/create.h b/src/create.h
index a994c84..98bb525 100644
--- a/src/create.h
+++ b/src/create.h
@@ -62,6 +62,7 @@ TextTypePtr CreateNewText (LayerTypePtr, FontTypePtr, LocationType,
 PolygonTypePtr CreateNewPolygon (LayerTypePtr, FlagType);
 PointTypePtr CreateNewPointInPolygon (PolygonTypePtr,
 				      LocationType, LocationType);
+void RefdesMapInit (void);
 ElementTypePtr CreateNewElement (DataTypePtr, ElementTypePtr,
 				 FontTypePtr, FlagType, char *, char *,
 				 char *, LocationType, LocationType, BYTE,
diff --git a/src/file.c b/src/file.c
index b6c67b9..653e1db 100644
--- a/src/file.c
+++ b/src/file.c
@@ -351,6 +351,9 @@ LoadPCB (char *Filename)
   PCBTypePtr newPCB = CreateNewPCB (False);
   Boolean units_mm;
 
+  /* Hack, load the mapping file */
+  RefdesMapInit ();
+
   /* new data isn't added to the undo list */
   if (!ParsePCB (newPCB, Filename))
     {
